modules/src/mock/context.rs

```diff
@@ -1152,10 +1220,13 @@ impl Ics18Context for MockContext {
 
     fn send(&mut self, msgs: Vec<Any>) -> Result<Vec<IbcEvent>, Ics18Error> {
         // Forward call to Ics26 delivery method.
-        let events = deliver(self, msgs).map_err(Ics18Error::transaction_failed)?;
-
+        let mut all_events = vec![];
+        for msg in msgs {
+            let (mut events, _) = deliver(self, msg).map_err(Ics18Error::transaction_failed)?;
+            all_events.append(&mut events);
+        }
```

Comment:
This could be changed to

```rust
let all_events = msgs
    .into_iter()
    .map(|msg| {
        let (events, _) = deliver(self, msg)?;
        Ok(events)
    })
    .collect::<Result<Vec<Vec<IbcEvent>>, _>>()
    .map_err(Ics18Error::transaction_failed)?
    .into_iter()
    .flatten()
    .collect();
```
I was not able to avoid calling `collect()` twice, which makes the implementation less efficient than this one... However, it's more idiomatic. If we can find a way to avoid the double `collect()`, then I think it would be better to apply the change. Otherwise, I'm okay if we leave it as is.

Reply:
Much of the heavy lifting is done in the first `collect()`, and makes use of [this `FromIterator` implementation](https://doc.rust-lang.org/std/result/enum.Result.html#impl-FromIterator%3CResult%3CA%2C%20E%3E%3E). Ideally, we could use this implementation by "collecting" into another iterator, flatten it, and then collect. However, AFAIK iterators don't implement the `FromIterator` trait.

There might be a clever way to use `flat_map` to achieve that as well.

Reply:
I figured out how to use `flat_map` in this context, but still with one quirk:
```rust
let all_events = msgs
            .into_iter()
            .flat_map(|msg| match deliver(self, msg) {
                Ok((events, _)) => events.into_iter().map(|event| Ok(event)).collect(),
                Err(e) => vec![Err(e)],
            })
            .collect::<Result<Vec<IbcEvent>, _>>()
            .map_err(Ics18Error::transaction_failed)?;
```
The last thing to figure out is how to properly have each `match` arm return the same iterator type, to avoid the extra `collect()` within `flat_map`.

Reply:
You can probably box them but it might require type annotations:

```rust
match deliver(self, msg) {
                Ok((events, _)) => Box::new(events.into_iter().map(|event| Ok(event))),
                Err(e) => Box::new(std::iter::once(Err(e))),
            }
```

Anyway, I feel like the original imperative implementation is a bit easier to read so I would suggest we stick with it, especially given that is only test code.